iree/hal/local/task_device.c

```diff
@@ -0,0 +1,331 @@
+  // buffers can contain inlined data uploads).
+  iree_arena_block_pool_t large_block_pool;
+
+  // iree_event_t pool for semaphore wait operations.
+  iree_hal_local_event_pool_t* event_pool;
+
+  iree_task_executor_t* executor;
+
+  iree_host_size_t loader_count;
+  iree_hal_executable_loader_t** loaders;
+
+  iree_host_size_t queue_count;
+  iree_hal_task_queue_t queues[];
+} iree_hal_task_device_t;
+
+static const iree_hal_device_vtable_t iree_hal_task_device_vtable;
+
+static iree_hal_task_device_t* iree_hal_task_device_cast(
+    iree_hal_device_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_task_device_vtable);
+  return (iree_hal_task_device_t*)base_value;
+}
+
+void iree_hal_task_device_params_initialize(
+    iree_hal_task_device_params_t* out_params) {
+  out_params->arena_block_size = 32 * 1024;
+  out_params->queue_count = 8;
+}
+
+static iree_status_t iree_hal_task_device_check_params(
+    const iree_hal_task_device_params_t* params) {
+  if (params->arena_block_size < 512) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "arena block size too small (< 512 bytes)");
+  }
+  if (params->queue_count == 0) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "at least one queue is required");
+  }
+  return iree_ok_status();
+}
+
+iree_status_t iree_hal_task_device_create(
+    iree_string_view_t identifier, const iree_hal_task_device_params_t* params,
+    iree_task_executor_t* executor, iree_host_size_t loader_count,
+    iree_hal_executable_loader_t** loaders, iree_allocator_t host_allocator,
+    iree_hal_device_t** out_device) {
+  IREE_ASSERT_ARGUMENT(params);
+  IREE_ASSERT_ARGUMENT(!loader_count || loaders);
+  IREE_ASSERT_ARGUMENT(out_device);
+  *out_device = NULL;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(z0,
+                                    iree_hal_task_device_check_params(params));
+
+  iree_hal_task_device_t* device = NULL;
+  iree_host_size_t total_size =
+      sizeof(*device) + params->queue_count * sizeof(*device->queues) +
+      identifier.size + loader_count * sizeof(*device->loaders);
+  iree_status_t status =
+      iree_allocator_malloc(host_allocator, total_size, (void**)&device);
+  if (iree_status_is_ok(status)) {
+    memset(device, 0, total_size);
+    iree_hal_resource_initialize(&iree_hal_task_device_vtable,
+                                 &device->base.resource);
+    iree_string_view_append_to_buffer(
+        identifier, &device->identifier,
+        (char*)device + sizeof(*device) +
+            params->queue_count * sizeof(*device->queues));
+    device->base.host_allocator = host_allocator;
+    iree_arena_block_pool_initialize(1024, host_allocator,
+                                     &device->small_block_pool);
+    iree_arena_block_pool_initialize(params->arena_block_size, host_allocator,
+                                     &device->large_block_pool);
```

Comment:
`arena_block_size` must be > `512` in `iree_hal_task_device_check_params`, but this makes it possible for `small_block_pool` to be larger than `large_block_pool`. How about increasing the min size of the large pool from `512` to `1024` above, or decreasing the size of the small pool from `1024` to `512` here?